src/NextGen/Framework/Receipts/DonationReceipt.php

```diff
@@ -0,0 +1,262 @@
+                ),
+                new ReceiptDetail(
+                    __('Donation Amount', 'give'),
+                    $this->donation->intendedAmount()->formatToDecimal()
+                ),
+            ]
+        );
+
+        if ($this->donation->feeAmountRecovered) {
+            $this->donationDetails->addDetail(
+                new ReceiptDetail(
+                    __('Processing Fee', 'give'),
+                    $this->donation->feeAmountRecovered->formatToDecimal()
+                )
+            );
+        }
+
+        $this->donationDetails->addDetail(
+            new ReceiptDetail(
+                __('Donation Total', 'give'),
+                $this->donation->amount->formatToDecimal()
+            )
+        );
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return void
+     */
+    private function fillDonorDetails()
+    {
+        $this->donorDetails->addDetails(
+            [
+                new ReceiptDetail(
+                    __('Donor Name', 'give'),
+                    trim("{$this->donation->firstName} {$this->donation->lastName}")
+                ),
+                new ReceiptDetail(
+                    __('Email Address', 'give'),
+                    $this->donation->email
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return void
+     */
+    private function fillAdditionalDetails()
+    {
+        if ($customFields = $this->getCustomFields()) {
+            $this->additionalDetails->addDetails($customFields);
+        }
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return void
+     */
+    private function fillSubscriptionDetails()
+    {
+        if ($this->donation->subscriptionId) {
+            $subscription = $this->donation->subscription;
+
+            $this->subscriptionDetails->addDetails([
+                new ReceiptDetail(
+                    __('Subscription', 'give'),
+                    sprintf(
+                        '%s / %s',
+                        $subscription->amount->formatToDecimal(),
+                        $subscription->period->getValue()
+                    )
+                ),
+                new ReceiptDetail(
+                    __('Subscription Status', 'give'),
+                    $subscription->status->getValue()
+                ),
+                new ReceiptDetail(
+                    __('Renewal Date', 'give'),
+                    $subscription->renewsAt->format('F j, Y')
+                ),
+                new ReceiptDetail(
+                    __('Progress', 'give'),
+                    sprintf(
+                        '%s / %s',
+                        count($subscription->donations),
+                        $subscription->installments > 0 ? $subscription->installments : __('Ongoing', 'give')
+                    )
+                ),
+            ]);
+        }
+    }
+
+    /**
+     * @unreleased
+     */
+    public function generate(): DonationReceipt
```

Comment:
Should we separate out the structure of a receipt and the filling of donation data? I'm thinking about extracting a `DonationReceiptBuilder` which fills the receipt using a donation. In which case, the lifecycle hook `givewp_donation_receipt_created` would pass a filled Receipt and the Donation.

```php
$receipt = new \Give\NextGen\Framework\Receipts\DonationReceipt();

do_action('givewp_donation_receipt_creating', $receipt);

// Fill the receipt using a donation.
$donationReceipt = DonationReceiptBuilder::build($receipt, $donation);

do_action('givewp_donation_receipt_created', $donationReceipt, $donation);
```

Reply:
I thought about that as well. The main reason I did not was because we would have to allow all the properties (`$donorDetails, $donationDetails`, etc.) to be mutated - where as only `$additionalDetails` currently has a setter.  I'm not entirely against that though and could be convinced otherwise if we think the separation is worth it.

Reply:
Okay, this is updated here 1d9832a0

After chatting and thinking through this, I decided it's okay for the different details to be mutable.  We can let the receipt object be pretty dumb in nature, just providing a shape really with entry points into its details.  Our donation receipt generator will mutate the receipt properties and fill it with details thats meaningful for a donation forms context.  This will make it flexible enough to be used in other contexts like we discussed with PDF receipts which would have its own builder to mutate a receipt object.

```